File: apps/go-idp/internal/auth/jwt.go
package auth

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
)

const AccessTokenTTLSeconds = 15 * 60

type AccessTokenClaims struct {
	Sub string `json:"sub"`
	Iat int64  `json:"iat"`
	Exp int64  `json:"exp"`
	Jti string `json:"jti"`
}

type JWTSigner struct {
	secret []byte
}

func NewJWTSigner(secret string) *JWTSigner {
	return &JWTSigner{secret: []byte(secret)}
}

func (s *JWTSigner) Sign(userID string) (token string, exp int64, jti string, err error) {
	iat := time.Now().Unix()
	exp = iat + AccessTokenTTLSeconds
	jti = uuid.NewString()
	t := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"sub": userID,
		"iat": iat,
		"exp": exp,
		"jti": jti,
	})
	signed, err := t.SignedString(s.secret)
	if err != nil {
		return "", 0, "", fmt.Errorf("sign jwt: %w", err)
	}
	return signed, exp, jti, nil
}

func (s *JWTSigner) Verify(token string) (*AccessTokenClaims, error) {
	parsed, err := jwt.Parse(token, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
		}
		return s.secret, nil
	})
	if err != nil {
		return nil, err
	}
	claims, ok := parsed.Claims.(jwt.MapClaims)
	if !ok || !parsed.Valid {
		return nil, errors.New("invalid claims")
	}

	out := &AccessTokenClaims{}
	if v, ok := claims["sub"].(string); ok {
		out.Sub = v
	}
	if v, ok := claims["iat"].(float64); ok {
		out.Iat = int64(v)
	}
	if v, ok := claims["exp"].(float64); ok {
		out.Exp = int64(v)
	}
	if v, ok := claims["jti"].(string); ok {
		out.Jti = v
	}
	if out.Sub == "" {
		return nil, errors.New("missing sub")
	}
	return out, nil
}

File: apps/node-idp/src/auth/jwt.ts
import { sign, verify } from 'hono/jwt';
import { config } from '../config';

export const ACCESS_TOKEN_TTL_SECONDS = 15 * 60;

export type AccessTokenClaims = {
  sub: string;
  iat: number;
  exp: number;
  jti: string;
};

export async function signAccessToken(userId: string): Promise<{ token: string; exp: number; jti: string }> {
  const iat = Math.floor(Date.now() / 1000);
  const exp = iat + ACCESS_TOKEN_TTL_SECONDS;
  const jti = crypto.randomUUID();
  const payload: AccessTokenClaims = { sub: userId, iat, exp, jti };
  const token = await sign(payload, config.jwtSecret, 'HS256');
  return { token, exp, jti };
}

export async function verifyAccessToken(token: string): Promise<AccessTokenClaims | null> {
  try {
    const decoded = await verify(token, config.jwtSecret, 'HS256');
    return decoded as AccessTokenClaims;
  } catch {
    return null;
  }
}
